### ISSUE TITLE
'spawn.before' is truncated when using a 'searchwindowsize'

### ISSUE THREAD

**Issue by alirez** [pexpect/pexpect#478]
When `searchwindowsize` is non-default, `spawn.before` no longer contains all the text up to the matched pattern.

The behavior seems to have changed after fd7332f5. Below is a quick demonstration of the issue.

The spawned shell prints out 500 lines and then the word PATTERN. It worked as expected before the above commit, i.e. 500 lines where in `.before` after the pattern matched.
```
$ git checkout fd7332f5~1 >& /dev/null
$ python
Python 2.7.13 (default, Nov 24 2017, 17:33:09)
[GCC 6.3.0 20170516] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pexpect
>>> child = pexpect.spawn('''/bin/bash -c "openssl rand -base64 {} | head -500 | nl --number-format=rz --number-width=5 2>&1 ; echo 'PATTERN'"'''.format(1024 * 1024 * 2), searchwindowsize=128)
>>> child.expect(['PATTERN'])
0
>>> print(len(child.before.splitlines()))
500
```

However, on fd7332f5:
```
$ git checkout fd7332f5 >& /dev/null
$ python
Python 2.7.13 (default, Nov 24 2017, 17:33:09)
[GCC 6.3.0 20170516] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pexpect
>>> child = pexpect.spawn('''/bin/bash -c "openssl rand -base64 {} | head -500 | nl --number-format=rz --number-width=5 2>&1 ; echo 'PATTERN'"'''.format(1024 * 1024 * 2), searchwindowsize=128)
>>> child.expect(['PATTERN'])
0
>>> print(len(child.before.splitlines()))
8
```
Only a fraction of the actual output is stored in `.before`

**Comment by takluyver** [comment]
@ryanpetrello could you take a look? It seems like this might be related to your PR #464. Thanks!

**Comment by ryanpetrello** [comment]
@takluyver this does look related to my change; the issue is that the new implementation throws away the anything beyond the buffer + search window on each chunk of data that's read, and as a result, `spawn.before` doesn't have the context it needs.  I'm going to work on a PR that uses an extra string/bytes buffer to track `before` content _up to_ a match.

**Comment by ryanpetrello** [comment]
@takluyver I think there's actually _another_ bug hidden here in the `after` attribute.  I'm going to write some tests and experiment.

**Comment by takluyver** [comment]
This should be fixed in #479
